### ISSUE TITLE
floor(x - S.Half).is_even returns True for negative even 'x'

### ISSUE THREAD

**Issue by aktech** [sympy/sympy#8853]
This should return False.
```python
In [7]: x =Symbol('x', even=True, negative=True)

In [8]: floor(x - S.Half).is_even
Out[8]: True
```

**Comment by smichr** [comment]
Hmmm...floor(x - S.Half) -> x. That's wrong.

**Comment by aktech** [comment]
Please note `x` is negative here:
So, for example:
```python
In [3]: floor(-2 - S.Half)
Out[3]: -3
```

**Comment by smichr** [comment]
Yes, so floor(x - S.Half) should be x - 1, not x.

**Comment by aktech** [comment]
Yes, that's what I meant, I thought you disagree.
:smile:

**Comment by smichr** [comment]
No -- we agree that floor(x-1/2) should be x - 1 if x is negative. So the attribute reported above is right...but it should have been False if x - 1 were returned from floor.

```
>>> var('x', even=True, negative=True)
x
>>> (x-1).is_even
False
```

**Comment by aktech** [comment]
Ok, got it.
It means the when (x-1) is returned from floor, the `even`assumptions are is wrong.

**Comment by smichr** [comment]
Yes.  Note the following in evalf.py

```
    # We can now easily find the nearest integer, but to find floor/ceil, we
    # must also calculate whether the difference to the nearest integer is
    # positive or negative (which may fail if very close).
```

That seems odd -- I would think that finding floor/ceil is easy but nearest integer the hard part. But it has something to do with being close to the middle (1/2) since the following do not fail to go the right way:

```
>>> var('x', even=True, negative=True)
x
>>> floor(x-Rational(1,3))
x - 1
>>> floor(x-Rational(2,3))
x - 1
```

**Comment by smichr** [comment]
Do you want to make a branch and try the following?

```
diff --git a/sympy/core/evalf.py b/sympy/core/evalf.py
index f84d372..49eb408 100644
--- a/sympy/core/evalf.py
+++ b/sympy/core/evalf.py
@@ -320,7 +320,7 @@ def get_integer_part(expr, no, options, return_ints=False):
     def calc_part(expr, nexpr):
         nint = int(to_int(nexpr, rnd))
         n, c, p, b = nexpr
-        if c != 1 and p != 0:
+        if (c != 1 or n == 1) and p != 0:
             expr = C.Add(expr, -nint, evaluate=False)
             x, _, x_acc, _ = evalf(expr, 10, options)
             try:

```

**Comment by aktech** [comment]
Yes sure, I am looking at that.

**Comment by aktech** [comment]
It works though:
```
In [3]: floor(x - S.Half).is_even
Out[3]: False
```
without breaking any test in test_eval.

**Comment by smichr** [comment]
So far so good! I can run a quick test of everything to see if this works and is ready for a PR.

**Comment by aktech** [comment]
That would be good.

**Comment by smichr** [comment]
The results are appearing [here](https://travis-ci.org/smichr/sympy/builds/47677120). Can you see that page? Only 3.4 is being run (and non-3.4 with special settings).

**Comment by aktech** [comment]
Yes, I can see that. Looks good so far.

**Comment by smichr** [comment]
Oops -- that's a different run. The correct one is [here](https://travis-ci.org/smichr/sympy/builds/47684323) and it has not yet begun to run.

**Comment by aktech** [comment]
Yes, got that, it's in the queue.

**Comment by aktech** [comment]
I think this is ready for a PR.
This needs to be merged, before merging [this](https://github.com/sympy/sympy/pull/8846).
Are you pushing? or shall I.

**Comment by smichr** [comment]
ceiling(pos_int + 1/2) had the same problem.

**Comment by aktech** [comment]
I have made a PR, for your changes.

**Comment by aktech** [comment]
> ceiling(pos_int + 1/2) had the same problem.

Ah! I can see, ceiling still have this problem.
```python
In [13]: x  = Symbol('x', even=True, positive=True)

In [14]: ceiling(x +S(1)/2).is_even
Out[14]: True

In [15]: ceiling(x + Rational(2,3)).is_even
Out[15]: False
```

**Comment by aktech** [comment]
Fixed here:  #8846
